iib/workers/tasks/build.py

```diff
@@ -337,6 +342,157 @@ def _get_resolved_image(pull_spec):
+                        'Index registry has not been initialized after '
+                        f'{conf.get("iib_grpc_max_tries")} tries'
+                    )
+                # not yet initialized, retry the loop
+                else:
+                    # sometimes, the registry service doesn't output any messages despite running
+                    # restarting it is an attempt to mitigate this issue
+                    rpc_proc.kill()
+                    continue
+            # process has terminated, something went wrong
+            else:
+                break
+        if registry_initialized:
+            break
+        else:
+            stderr = rpc_proc.stderr.read()
+            if 'address already in use' in stderr:
+                if i == conf['iib_grpc_max_port_tries'] - 1:
+                    err_msg = (
+                        'No free port has been found after '
+                        f'{conf.get("iib_grpc_max_port_tries")} attempts.'
+                    )
+                    log.error(err_msg)
+                    raise IIBError(err_msg)
+                log.info('Port %d is in use, trying another.', port)
+                port += 1
+            else:
+                raise IIBError(f'Command "{" ".join(cmd)}" has failed with error "{stderr}"')
+
+    return (port, rpc_proc)
+
+
+@retry(attempts=2, wait_on=IIBError, logger=log)
+def _get_present_bundles(from_index, base_dir):
+    """
+    Get a list of bundles already present in the index image.
+
+    :param str from_index: index image to inspect.
+    :param str base_dir: base directory to create temporary files in.
+    :return: list of present bundles as provided by the grpc query.
+    :rtype: list
+    :raises IIBError: if any of the commands fail.
+    """
+    db_path = _get_index_database(from_index, base_dir)
+    port, rpc_proc = _serve_index_registry(db_path)
+
+    bundles = run_cmd(
+        ['grpcurl', '-plaintext', f'localhost:{port}', 'api.Registry/ListBundles'],
+        exc_msg='Failed to get bundle data from index image',
+    )
+    rpc_proc.kill()
+
+    # Transform returned data to parsable json
+    present_bundles = [json.loads(bundle) for bundle in re.split(r'(?<=})\n(?={)', bundles)]
+    return present_bundles
+
+
+@retry(attempts=2, wait_on=IIBError, logger=log)
+def _get_missing_bundles(present_bundles, bundles):
+    """
+    Filter out bundles to only those not present in the index image.
+
+    :param list present_bundles: list of bundles present in the index image, as provided by opm.
+    :param list bundles: bundles requested to be added to the index image.
+    :return: list of bundles not present in the index image.
+    :rtype: list
+    """
+    # we only need package name and version
+    present_bundle_ids = []
+    present_bundle_ids_path = []
+    present_bundle_hashes = []
+    for bundle in present_bundles:
+        present_bundle_ids.append((bundle['packageName'], bundle['version']))
```

Comment:
I am sorry I am a bit confused with `present_bundle_ids` here. Why can't we only use `present_bundle_hashes` and `present_bundle_ids_path`? 

IIUC, `present_bundle_ids_path` is going to be (Name, version-release) which should be sufficient to compare in the second for loop you have to go over the bundles submitted by the user. Am I missing something here?

Reply:
Basically, 'bundlePath' entry provided by opm will contain bundle specified by digest about 99% of the time (as this is how it was added to the index image). On the other hand, nothing prevents it being specified by tag, so I wanted to account for that possibility as well.

During prototyping and testing I found out that the data matching might still result in some false negatives, so I wanted to make the process as robust as possible. This is why i compare both name+version and digests. Unfortunately, neither name+version difference nor digest difference guarantee that opm won't give unique constraint error. But I should note that I tested this on stage data, and in production these inconsistencies will be far less likely.

Reply:
Correct, I get the point of checking both NVR and the sha but I don't think you have to keep a third list of present_bundle_ids which is essentially only NV. You will anyways check NVR when you compare the list with `present_bundle_ids_path`. So I think you can simply remove the parts with `present_bundle_ids`. Unless I am missing something.

Reply:
As I've said, I encountered some divergence in data provided by grpc queries and actual index image data. For example, grpc said bundle package name + version were 'bundle:1.0', while in reality they were saved in the index image as 'bundle:1.0.5'. This ccaused a unique constraint error, even though my script hasn't detected a match. I don't know if this is likely (or possible) to happen in production data, but I thought that increasing redundancy will reduce the probability of these inconsistencies.

So, I check for a match in packagename+version, then if bundle was specified by digest (most cases), I check for a match in digests, and if it was specified by tag then I check for name+version again from a different source.